tests/test_layouts.py

```diff
@@ -250,6 +251,74 @@ def testMDS(self):
         lo = g.layout("mds")
         self.assertTrue(isinstance(lo, Layout))
 
+    def testUMAP(self):
+        g = Graph()
+
+        # These distinct types of exceptions look ugly... can we improve?
+        self.assertRaises(
+                InternalError, g.layout_umap,
+                min_dist=-0.01,
+            )
+
+        self.assertRaises(
+                ValueError, g.layout_umap,
+                epochs=-1,
+            )
+
+        self.assertRaises(
+                InternalError, g.layout_umap,
+                sampling_prob=-0.01,
+            )
+
+        self.assertRaises(
+                InternalError, g.layout_umap,
+                sampling_prob=1.01,
+            )
+
+        # Empty graph
+        lo = g.layout_umap()
+        self.assertTrue(isinstance(lo, Layout))
+        self.assertEqual(lo.coords, [])
+
+        # Singleton graph
+        g = Graph(n=1)
+        lo = g.layout_umap()
+        self.assertEqual(lo.coords, [[0, 0]])
+
+        # Graph with two articulation points
+        edges = [
+            0, 1, 0, 2, 0, 3, 1, 2, 1, 3, 2, 3,
+            3, 4, 4, 5, 5, 6,
+            6, 7, 7, 8, 6, 8, 7, 9, 6, 9, 8, 9, 7, 10, 8, 10, 9, 10,
+            10, 11, 9, 11, 8, 11, 7, 11,
+            ]
+        edges = list(zip(edges[::2], edges[1::2]))
+        dist = [
+            0.1, 0.09, 0.12, 0.09, 0.1, 0.1,
+            0.9, 0.9, 0.9,
+            0.2, 0.1, 0.1, 0.1, 0.1, 0.1, 0.08, 0.05, 0.1, 0.08, 0.12, 0.09, 0.11
+            ]
+        g = Graph(edges)
+        lo = g.layout_umap(dist=dist, epochs=500, sampling_prob=0.3)
+        self.assertTrue(isinstance(lo, Layout))
+
+        # One should get two clusters in this case
+        xmax = max([x for (x, y) in lo.coords])
+        ymax = max([y for (x, y) in lo.coords])
+        xmin = min([x for (x, y) in lo.coords])
+        ymin = min([y for (x, y) in lo.coords])
+        distmax = max(xmax - xmin, ymax - ymin)
+        for iclu in range(0, 8, 7):
+            xclu = sum([x for (x, y) in lo.coords[iclu:iclu + 4]]) / 4
+            yclu = sum([y for (x, y) in lo.coords[iclu:iclu + 4]]) / 4
+            for i in range(4):
+                dx = lo.coords[iclu + i][0] - xclu
+                dy = lo.coords[iclu + i][1] - yclu
+                dist = sqrt(dx * dx + dy * dy)
```

Comment:
Use `math.hypot()` because that is less susceptible to overflows (and it also works in higher dimensions).